### PR TITLE
Ref/Span fixes

### PR DESCRIPTION
**Pull request by TIHan**
This is intended to fix several safety rules with ref returns and span.

### CODE REVIEW

File: src/fsharp/PostInferenceChecks.fs [hunk 1]
```diff
@@ -741,28 +773,54 @@ and CheckExpr (cenv:cenv) (env:env) origExpr (context:PermitByRefExpr) : bool =
...
 
     // Check an application
     | Expr.App(f,_fty,tyargs,argsl,m) ->
 
         CheckTypeInstNoByrefs cenv env m tyargs
         CheckExprNoByrefs cenv env f
 
+        let contexts = mkArgsForAppliedExpr false argsl f
+
+        let _hasHeapSpanLikeArg, _hasStackSpanLikeArg, hasStackByRefLikeArg =
```
Comment (TIHan) [inline src/fsharp/PostInferenceChecks.fs]: This code is a bit complicated, but something like this is necessary when we start supporting stack referring span likes, and mix call functions with arguments that have both stack + heap referring span likes.

File: src/fsharp/PostInferenceChecks.fs [hunk 2]
```diff
@@ -887,35 +945,39 @@ and CheckExprOp cenv env (op,tyargs,args,m) context expr =
...
                 errorR(Error(FSComp.SR.chkNoByrefAddressOfLocal(vref.DisplayName), m))
 
-        let limit1 = IsLimited cenv env m vref
-        let limit2 = CheckExprsNoByRefLike cenv env args                   
-        limit1 || limit2
+        let limit1 = 
+            if IsValLocalAndNotArgument env vref.Deref then
+                LimitValFlags.StackReferringByRef
+            else
+                LimitValFlags.None
```
Comment (TIHan) [inline src/fsharp/PostInferenceChecks.fs]: Hmm, at this point I may want to return a LimitValFlags.ByRef here.

**Comment by TIHan** [comment]
I'm using these as tests:

As of this commit: https://github.com/Microsoft/visualfsharp/pull/5146/commits/1359ca717feecf67b0d23b3482275a5f8819d8c0

```fsharp
let mutable beef = 1

let returning_span (x: byref<int>) = Span.Empty

let returning_ref (x: byref<int>) = &x

let returning_ref2 (x: byref<int>) (y: byref<int>) = &x

let should_not_work1 () =
    let mutable x = 1
    &x

let should_not_work2 () =
    let mutable x = 1
    let mutable x = (let mutable y = &x in &y)

    let y = &returning_ref &x
    &y

let should_not_work3 () =
    let mutable x = 1
    let mutable x =
        match &x with
        | x -> &x

    let y = &returning_ref &x
    &y

let should_not_work4 () =
    let mutable x = 1
    let mutable y = &x
    y <- 1
    &y

let should_not_work5 () =
    let x = 1
    let y = &x
    &y

let should_not_work6 () =
    let mutable x = 1
    &returning_ref &x

let should_not_work7 () =
    let mutable x = 1
    &returning_ref2 &beef &x

let should_not_work8 () =
    let mutable x = 1
    let y = &returning_ref2 &beef &x
    &y

let should_work1 () =
    Span.Empty

let should_work3 () =
    let y = &returning_ref &beef
    &y
    
let should_work2 () =
    let x = Span.Empty
    x

let should_work4 () =
    &returning_ref2 &beef &beef

let should_work5 () =
    let mutable x = 1
    returning_ref2 &x &x

let should_work6 () =
    let mutable x = 1
    returning_span &x

let should_work7 () =
    let mutable x = 1
    let y = returning_span &x
    y
```

All behave as intended except for "should_not_work2" and "should_not_work3", which pass in the compiler. Will need to dig into the ast a bit more.

**Comment by cartermp** [comment]
Tagging as 15.8 - we're in escrow for Preview 3, so ideally this would get in there, but we may have to get it into Preview 4 instead. Which is probably fine, it would just mean that we'd have a set of known issues for Span in Preview 3.

**Comment by TIHan** [comment]
Here is some updated cases:

```fsharp
let mutable beef = 1

let returning_span (x: byref<int>) = Span.Empty

let returning_ref (x: byref<int>) = &x

let returning_ref2 (x: byref<int>) (y: byref<int>) = &x

let passing_span_returning_span (x: Span<int>) = Span.Empty

let passing_span_returnin_ref (x: Span<int>) = &x.[0]

let passing_span2 (x: Span<int>) (y: Span<int>) = ()

let passing_byref_of_span_returning_span (x: byref<Span<int>>) = x

let passing_byref_of_span_returning_byref_of_span (x: byref<Span<int>>) = &x

let should_not_work () =
    let test = ReadOnlySpan.Empty
    let f = fun () -> test
    ()

let should_not_work1 () =
    let mutable x = 1
    &x

let should_not_work2 () =
    let mutable x = 1
    let mutable x = (let mutable y = &x in &y)

    let y = &returning_ref &x
    &y

let should_not_work3 () =
    let mutable x = 1
    let mutable x =
        match &x with
        | x -> &x

    let y = &returning_ref &x
    &y

let should_not_work4 () =
    let mutable x = 1
    let mutable y = &x
    y <- 1
    &y

let should_not_work5 () =
    let x = 1
    let y = &x
    &y

let should_not_work6 () =
    let mutable x = 1
    &returning_ref &x

let should_not_work7 () =
    let mutable x = 1
    &returning_ref2 &beef &x

let should_not_work8 () =
    let mutable x = 1
    let y = &returning_ref2 &beef &x
    &y

let should_not_work9 () =
    let mutable s = Span.Empty
    passing_span_returning_span s

let should_not_work10 () =
    let mutable s = Span.Empty
    &passing_span_returnin_ref s

let should_not_work11 () =
    let mutable s = Span.Empty
    &passing_span_returnin_ref (passing_span_returning_span s)

let should_not_work12 () =
    let mutable s = Span.Empty
    passing_byref_of_span_returning_span &s

let should_not_work13 () =
    let mutable s = Span.Empty
    s

let should_not_work14 () =
    let mutable s = Span.Empty
    let mutable x = &s
    passing_byref_of_span_returning_span &x

let should_work1 () =
    Span.Empty

let should_work3 () =
    let y = &returning_ref &beef
    &y
    
let should_work2 () =
    let x = Span.Empty
    x

let should_work4 () =
    &returning_ref2 &beef &beef

let should_work5 () =
    let mutable x = 1
    returning_ref2 &x &x

let should_work6 () =
    let mutable x = 1
    returning_span &x

let should_work7 () =
    let mutable x = 1
    let y = returning_span &x
    y

let should_work8 () =
    let s = Span.Empty
    passing_span_returning_span s

let should_work9 () =
    let mutable s = Span.Empty
    passing_span_returnin_ref s

let should_work10 () =
    let mutable s = Span.Empty
    passing_span_returnin_ref (passing_span_returning_span s)

let should_work11 () =
    let s = Span.Empty
    &passing_span_returnin_ref (passing_span_returning_span s)

let should_work12 () =
    let s = Span.Empty
    passing_byref_of_span_returning_span &s

let should_work13 () =
    let s = Span.Empty
    let mutable x = &s
    passing_byref_of_span_returning_span &x

let should_work14 (s: byref<Span<int>>) =
    passing_byref_of_span_returning_byref_of_span &s

let should_work15 (s: byref<Span<int>>) =
    let x = &s
    let y = &passing_byref_of_span_returning_byref_of_span &x
    &y

type TestDelegate = delegate of byref<Span<int>> -> Span<int>

let should_work16 (s: byref<Span<int>>) =
    let f = TestDelegate(fun _ -> Span.Empty)
    //let mutable s = Span.Empty
    f.Invoke(&s)

type IBeef =

    abstract Test : byref<Span<int>> -> byref<Span<int>>

let should_work17 () =
    { new IBeef with
        member __.Test s = &s
    }

```

**Comment by TIHan** [comment]
All tests are passing, this is a good sign. I will add more tests, code cleanup and comments.

**Comment by TIHan** [comment]
I'm going to write more tests. Going to port some roslyn ones over.

**Comment by forki** [comment]
This is starting to become an epic - shouldn't we try to get it into the bank now and then you can send new PRs on top?

**Comment by TIHan** [comment]
We set a plan to get this in next week, hopefully Monday if all goes well. All that remains are a few more tests and I think it should be good to go.

**Comment by TIHan** [comment]
test Ubuntu16.04 Release_fcs Build please

**Comment by TIHan** [comment]
test Windows_NT Release_fcs Build please

**Comment by TIHan** [comment]
test Windows_NT Release_ci_part4 Build please

**Comment by forki** [comment]
Yay

Will Smith <notifications@github.com> schrieb am Mi., 20. Juni 2018, 21:13:

**Pull request merged**